pkg/bucket/lifecycle/expiration.go

```diff
@@ -61,73 +61,73 @@ type ExpirationDate struct {
 	time.Time
 }
 
-// UnmarshalXML parses date from Expiration and validates date format
-func (eDate *ExpirationDate) UnmarshalXML(d *xml.Decoder, startElement xml.StartElement) error {
-	var dateStr string
-	err := d.DecodeElement(&dateStr, &startElement)
-	if err != nil {
+// MarshalXML encodes to XML data.
+func (edate ExpirationDate) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
+	if edate.IsZero() {
+		return errLifecycleInvalidDate
+	}
+
+	expected := time.Date(edate.Year(), edate.Month(), edate.Day(), 0, 0, 0, 0, time.UTC)
+	if !edate.Equal(expected) {
+		return errLifecycleDateNotMidnight
+	}
+
+	return e.EncodeElement(edate.Format(time.RFC3339), start)
+}
+
+// UnmarshalXML decodes XML data.
+func (edate *ExpirationDate) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	var s string
+	if err := d.DecodeElement(&s, &start); err != nil {
 		return err
 	}
+
 	// While AWS documentation mentions that the date specified
 	// must be present in ISO 8601 format, in reality they allow
 	// users to provide RFC 3339 compliant dates.
-	expDate, err := time.Parse(time.RFC3339, dateStr)
+	date, err := time.Parse(time.RFC3339, s)
 	if err != nil {
 		return errLifecycleInvalidDate
 	}
-	// Allow only date timestamp specifying midnight GMT
-	hr, min, sec := expDate.Clock()
-	nsec := expDate.Nanosecond()
-	loc := expDate.Location()
-	if !(hr == 0 && min == 0 && sec == 0 && nsec == 0 && loc.String() == time.UTC.String()) {
+
+	expected := time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, time.UTC)
+	if !date.Equal(expected) {
 		return errLifecycleDateNotMidnight
 	}
 
-	*eDate = ExpirationDate{expDate}
+	*edate = ExpirationDate{date}
 	return nil
 }
 
-// MarshalXML encodes expiration date if it is non-zero and encodes
-// empty string otherwise
-func (eDate *ExpirationDate) MarshalXML(e *xml.Encoder, startElement xml.StartElement) error {
-	if *eDate == (ExpirationDate{time.Time{}}) {
-		return nil
-	}
-	return e.EncodeElement(eDate.Format(time.RFC3339), startElement)
-}
-
 // Expiration - expiration actions for a rule in lifecycle configuration.
 type Expiration struct {
-	XMLName xml.Name       `xml:"Expiration"`
-	Days    ExpirationDays `xml:"Days,omitempty"`
-	Date    ExpirationDate `xml:"Date,omitempty"`
+	XMLName xml.Name        `xml:"Expiration"`
+	Days    *ExpirationDays `xml:"Days,omitempty"`
```

Comment:
We should avoid really pointer-based variables here, that was the purpose of the previous code to ensure that we do not use unnecessary pointer-based approach when the variables which can have safe default values. 

With this, we are going against the Go's idiomatic approach for simpletons, using pointers for composite data structures it makes sense as we cannot validate the individual fields easily. 

An `int`, `string` shouldn't have such a requirement in the first place using pointers for that is like bringing Java-style here. 

This portion is preferred to be in the previous style as its easier to read and understand, otherwise, there is too much safety code that needs to be added elsewhere before accessing any of these fields.

Reply:
this is the prime reason why aws-sdk-go is disliked in the Go community for being terrible at creating consumable APIs and also going against the interest of the programming language. 

A prime example for that is `aws.String("mystring")` converts a string into a pointer if you pass `""` returns nil